<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
<title>GenomeTools</title>
<meta http-equiv="content-type" content="text/html; charset=iso-8859-1">
<meta name="description" content="The GenomeTools genome analysis system is a free collection of bioinformatics tools (in the realm of genome informatics) combined into a single binary.">
<meta name="keywords" content="genome analysis, GFF3, genome annotation,
annotation drawing, AnnotationSketch, LTR prediction, bioinformatics, computational biology">
<link rel="stylesheet" type="text/css" href="style.css">
</head>
<body>
  <div id="menu">
    <ul>
      <li><a id="current" href="index.html">Overview</a></li>
      <li><a href="https://github.com/genometools/genometools/releases">Releases</a></li>
      <li><a href="pub/">Archive</a></li>
      <li><a href="https://github.com/genometools/genometools">Browse source</a></li>
      <li><a href="http://github.com/genometools/genometools/issues/">Issue tracker</a></li>
      <li><a href="documentation.html">Documentation</a></li>
      <li><a href="annotationsketch.html"><em>AnnotationSketch</em></a></li>
      <li><a href="/cgi-bin/gff3validator.cgi">GFF3 validator</a></li>
      <li><a href="license.html">License</a></li>
    </ul>
  </div>
  <div id="main">
    <h1><i>GenomeTools</i></h1>
    <h3>The versatile open source genome analysis software</h3>
    <p id="examplesketch" >
      <a href="annotationsketch.html">
        <img src="images/annotation.png" alt="[Genome Annotation]">
      </a>
    </p>
    <p>
    The <i>GenomeTools</i> genome analysis system is a
    <a href="license.html">free</a> collection of bioinformatics
    <a href="tools.html">tools</a> (in the realm of
    genome informatics) combined into a single binary named <em>gt</em>.
    It is based on a C library named &ldquo;libgenometools&rdquo; which consists of
    several modules.
    </p>
    <p>
    If you are interested in gene prediction, have a look at
    <a href="http://genomethreader.org" title="GenomeThreader gene prediction
      software"><i>GenomeThreader</i></a>.
    </p>
    <h2>Platforms</h2>
    <p>
    <i>GenomeTools</i> has been designed to run on every
    <a href="http://en.wikipedia.org/wiki/POSIX">POSIX</a> compliant
    <a href="http://en.wikipedia.org/wiki/UNIX">UNIX</a> system &mdash; for example Linux or
    macOS. A reduced Windows version is also available. Please report problems to our
    <a href="http://github.com/genometools/genometools/issues/">issue tracker</a>.
    </p>
    <h2><em>GenomeTools</em> publication</h2>
      <div class="ref">
       G. Gremme, S. Steinbiss and S. Kurtz.<br>
       <i>GenomeTools</i>: a comprehensive software library for efficient
       processing of structured genome annotations. <br>
       <a href="http://doi.ieeecomputersociety.org/10.1109/TCBB.2013.68">
         IEEE/ACM Transactions on Computational Biology and Bioinformatics 2013, 10(3):645&ndash;656
       </a>
      </div>
    <h2>Published software in <em>GenomeTools</em></h2>
    <p>
    The <em>GenomeTools</em> distribution includes several published software
    tools:
    </p>
    <ul class="pubs">
      <li><em><a href="https://www.zbh.uni-hamburg.de/forschung/gi/software/ltrharvest.html">LTRharvest</a></em>, an efficient and flexible software tool for
      <em>de novo</em> detection of LTR retrotransposons.
      <div class="ref">
       D. Ellinghaus, S. Kurtz, and U. Willhoeft.<br>
       <i>LTRharvest</i>, a efficient and flexible software for <i> de novo </i>
       detection of LTR retrotransposons. <br>
       <a href="http://www.biomedcentral.com/1471-2105/9/18/abstract">
         BMC Bioinformatics 2008, 9:18
       </a>
      </div>
      </li>
      <li><em><a href="https://www.zbh.uni-hamburg.de/forschung/gi/software/tallymer.html">Tallymer</a></em>, a collection of flexible and memory-efficient
      programs for <em>k</em>-mer counting and indexing of large sequence sets.
      <div class="ref">
       S. Kurtz, A. Narechania, J.C. Stein, and D. Ware.<br>
       A new method to compute K-mer frequencies and its application to annotate
        large repetitive plant genomes.<br>
       <a href="http://www.biomedcentral.com/1471-2164/9/517/abstract">
         BMC Genomics 2008, 9:517
       </a>
      </div>
      </li>
      <li><tt>uniquesub</tt>, a program for computing minimum unique substrings.
      <div class="ref">
       S. Gr&auml;f, F.G.G. Nielsen, S. Kurtz, M.A. Huynen, E. Birney,
       H. Stunnenberg, and P. Flicek.<br>
       Optimized design and assessment of whole genome tiling arrays.<br>
       <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/23/13/i195">
         Bioinformatics 2007, 23(13):i195&ndash;i204
       </a>
      </div>
      </li>
      <li><em>AnnotationSketch</em>, a library for drawing genome annotations.
      <div class="ref">
       S. Steinbiss, G. Gremme, C. Sch&auml;rfer, M. Mader and S. Kurtz.<br>
       <em>AnnotationSketch</em>: a genome annotation drawing library.<br>
       <a href="http://bioinformatics.oxfordjournals.org/cgi/content/short/25/4/533">
         Bioinformatics 2009, 25(4):533&ndash;534
       </a>
      </div>
      </li>
      <li><em><a href="https://www.zbh.uni-hamburg.de/forschung/gi/software/ltrdigest.html">LTRdigest</a></em>, a software tool for automated annotation of internal features of LTR retrotransposons.
      <div class="ref">
       S. Steinbiss, U. Willhoeft, G. Gremme and S. Kurtz.<br>
       Fine-grained annotation and classification of <em>de novo</em> predicted LTR retrotransposons.<br>
       <a href="http://nar.oxfordjournals.org/cgi/content/full/37/21/7002">
         Nucleic Acids Research 2009, 37(21):7002&ndash;7013
       </a>
      </div>
      </li>
      <li><em>MetaGenomeThreader</em>, a software to predict genes, such as PCS's (predicted coding sequences) in sequences of metagenome projects.
      <div class="ref">
       D.J. Schmitz-H&uuml;bsch and S. Kurtz.<br>
       MetaGenomeThreader: A software tool for predicting genes in DNA-sequences of metagenome projects.<br>
       In R. Daniel and W. Streit (Eds.), <em>Metagenomics. Methods in Molecular Biology</em>, 325&ndash;338, Humana Press, Totowa, NJ, ISBN 978-1-60761-822-5
      </div>
      </li>
      <li><i><a href="https://www.zbh.uni-hamburg.de/forschung/gi/software/gtencseq.html">GtEncseq</a></i>, a compressed biosequence representation with many features.
      <div class="ref">
       S. Steinbiss and S. Kurtz.<br>
       A New Efficient Data Structure for Storage and Retrieval of Multiple Biosequences.<br>
       <a href="http://doi.ieeecomputersociety.org/10.1109/TCBB.2011.146">
         IEEE/ACM Transactions on Computational Biology and Bioinformatics 2012, 9(2):345&ndash;357
       </a>
      </div>
      </li>
      <li><i><a href="https://www.zbh.uni-hamburg.de/forschung/gi/software/readjoiner.html">Readjoiner</a></i>, a sequence assembler based on the assembly string graph framework.
      <div class="ref">
       G. Gonnella and S. Kurtz.<br>
       Readjoiner: a fast and memory efficient string graph-based sequence assembler.<br>
       <a href="http://www.biomedcentral.com/1471-2105/13/82">BMC Bioinformatics 2012, 13:82</a>
      </div>
      </li>
    </ul>
    <p>
    The following software tools are based on the <em>GenomeTools</em> library:
    </p>
    <ul class="pubs">
      <li><i><a href="http://genomethreader.org">GenomeThreader</a></i>,
        a software tool to compute eukaryote gene structure predictions.
      <div class="ref">
        G. Gremme, V. Brendel, M.E. Sparks, and S. Kurtz.<br>
        Engineering a software tool for gene structure prediction in higher organisms. <br>
        <a href="http://genomethreader.org/doc/GreBreSpaKur2005.pdf">
          Information and Software Technology 2005, 47(15):965&ndash;978
        </a>
      </div>
      </li>
      <li><em><a href="https://www.zbh.uni-hamburg.de/forschung/gi/software/fish-oracle-2.html">FISH Oracle</a></em>,
        a web server for visualizing cancer genomics data.
      <div class="ref">
       M. Mader, R. Simon, S. Steinbiss and S. Kurtz.<br>
       FISH Oracle: a web server for flexible visualization of DNA copy number data in a genomic context. <br>
       <a href="http://www.jclinbioinformatics.com/content/1/1/20">
         Journal of Clinical Bioinformatics 2011, 1:20
       </a>
      </div>
      </li>
      <li><em><a href="https://parseval.sourceforge.net">ParsEval</a></em>, a
      tool for comparing genome annotations.
      <div class="ref">
       D.S. Standage and V.P. Brendel.<br>
       ParsEval: parallel comparison and analysis of gene structure annotations. <br>
       <a href="http://www.biomedcentral.com/1471-2105/13/187">
         BMC Bioinformatics 2012, 13:187
       </a>
      </div>
      </li>
      <li><em><a href="https://www.zbh.uni-hamburg.de/forschung/gi/software/ltrsift.html">LTRsift</a></em>,
        a graphical interface for working with LTR retrotransposon predictions.
      <div class="ref">
       S. Steinbiss, S. Kastens and S. Kurtz.<br>
       LTRsift: a graphical user interface for semi-automatic classification
       and postprocessing of de novo detected LTR retrotransposons. <br>
       <a href="http://www.mobilednajournal.com/content/3/1/18">
         Mobile DNA 2012, 3:18
       </a>
      </div>
      </li>
      <li><em><a href="https://www.zbh.uni-hamburg.de/forschung/gi/software/fish-oracle-2.html">FISH Oracle 2</a></em>,
      a web-based software tool for the integrative analysis of cancer genomics data.
      <div class="ref">
       M. Mader, R. Simon, and S. Kurtz.<br>
       FISH Oracle 2: a web server for integrative visualization of genomic data
       in cancer research. <br>
       <a href="http://www.jclinbioinformatics.com/content/4/1/5">
         Journal of Clinical Bioinformatics 2014, 4:5
       </a>
      </div>
      </li>
    </ul>
    <h2>Contact</h2>
    <p>
    The <em>GenomeTools</em> are being developed by: <br>
    Gordon Gremme, Sascha Steinbiss, Stefan Kurtz and <a href="https://github.com/genometools/genometools/blob/master/CONTRIBUTORS">others</a><br>
    <a href="https://www.zbh.uni-hamburg.de/en/forschung/gi.html">Genome Informatics Research Group</a><br>
     <a href="https://www.zbh.uni-hamburg.de">Center for Bioinformatics, University of Hamburg</a>
    </p>
<div id="footer">
Copyright &copy; 2006-2023
The <i>GenomeTools</i> authors. Last update: 2014-11-05
</div>
</div>
</body>
</html>
