<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="application/xhtml+xml; charset=UTF-8" />
<title>GenomeTools - manual page for GT-ENCSEQ-ENCODE(1)</title>
<link rel="stylesheet" type="text/css" href="style.css">
<link rel="stylesheet" href="..//style.css" type="text/css" />
</head>
<body>
<div id="menu">
<ul>
<li><a href="../index.html">Overview</a></li>
<li><a href="https://github.com/genometools/genometools/releases">Releases</a></li>
<li><a href="pub/">Archive</a></li>
<li><a href="https://github.com/genometools/genometools">Browse source</a></li>
<li><a href="http://github.com/genometools/genometools/issues/">Issue tracker</a></li>
<li><a href="../documentation.html">Documentation</a></li>
  <ul class="submenu">
    <li><a id="current" href="../tools.html">Tools</a></li>
    <li><a href="../manuals.html">Manuals</a></li>
    <li><a href="../libgenometools.html">C API</a></li>
    <li><a href="../docs.html"><tt>gtscript</tt> docs</a></li>
    <li><a href="../contract.html">Development Contract</a></li>
    <li><a href="../contribute.html">Contribute</a></li>
  </ul>
<li><a href="../annotationsketch.html"><tt>AnnotationSketch</tt></a></li>
<li><a href="../cgi-bin/gff3validator.cgi">GFF3 validator</a></li>
<li><a href="../license.html">License</a></li>
</ul>
</div>
<div id="main">
<div class="sect1">
<h2 id="_name">NAME</h2>
<div class="sectionbody">
<div class="paragraph"><p>gt-encseq-encode - Encode sequence files (FASTA/FASTQ, GenBank, EMBL) efficiently.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="_synopsis">SYNOPSIS</h2>
<div class="sectionbody">
<div class="paragraph"><p><strong>gt encseq encode</strong> sequence_file [sequence_file [sequence_file &#8230;]]</p></div>
</div>
</div>
<div class="sect1">
<h2 id="_description">DESCRIPTION</h2>
<div class="sectionbody">
<div class="dlist"><dl>
<dt class="hdlist1">
<strong>-showstats</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
show compression results (default: no)
</p>
</dd>
<dt class="hdlist1">
<strong>-ssp</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
output sequence separator positions to file (default: yes)
</p>
</dd>
<dt class="hdlist1">
<strong>-des</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
output sequence descriptions to file (default: yes)
</p>
</dd>
<dt class="hdlist1">
<strong>-sds</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
output sequence description separator positions to file (default: yes)
</p>
</dd>
<dt class="hdlist1">
<strong>-md5</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
output MD5 sums to file (default: yes)
</p>
</dd>
<dt class="hdlist1">
<strong>-clipdesc</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
clip descriptions after first whitespace (default: no)
</p>
</dd>
<dt class="hdlist1">
<strong>-sat</strong> [<em>string</em>]
</dt>
<dd>
<p>
specify kind of sequence representation
by one of the keywords direct, bytecompress, eqlen, bit, uchar, ushort, uint32 (default: undefined)
</p>
</dd>
<dt class="hdlist1">
<strong>-dna</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
input is DNA sequence (default: no)
</p>
</dd>
<dt class="hdlist1">
<strong>-protein</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
input is protein sequence (default: no)
</p>
</dd>
<dt class="hdlist1">
<strong>-plain</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
process as plain text (default: no)
</p>
</dd>
<dt class="hdlist1">
<strong>-dust</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
mask low-complexity regions using the dust algorithm (default: no)
</p>
</dd>
<dt class="hdlist1">
<strong>-dustwindow</strong> [<em>value</em>]
</dt>
<dd>
<p>
windowsize for the dust algorithm (default: 64)
</p>
</dd>
<dt class="hdlist1">
<strong>-dustthreshold</strong> [<em>value</em>]
</dt>
<dd>
<p>
threshold for the dust algorithm (default: 2.000000)
</p>
</dd>
<dt class="hdlist1">
<strong>-dustlink</strong> [<em>value</em>]
</dt>
<dd>
<p>
Max. distance between regions masked by dust before merging. (default: 1)
</p>
</dd>
<dt class="hdlist1">
<strong>-indexname</strong> [<em>string</em>]
</dt>
<dd>
<p>
specify name for index to be generated (default: undefined)
</p>
</dd>
<dt class="hdlist1">
<strong>-smap</strong> [<em>string</em>]
</dt>
<dd>
<p>
specify file containing a symbol mapping (default: undefined)
</p>
</dd>
<dt class="hdlist1">
<strong>-lossless</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
allow lossless original sequence retrieval (default: no)
</p>
</dd>
<dt class="hdlist1">
<strong>-v</strong> [<em>yes|no</em>]
</dt>
<dd>
<p>
be verbose (default: no)
</p>
</dd>
<dt class="hdlist1">
<strong>-help</strong> 
</dt>
<dd>
<p>
display help for basic options and exit
</p>
</dd>
<dt class="hdlist1">
<strong>-help+</strong> 
</dt>
<dd>
<p>
display help for all options and exit
</p>
</dd>
<dt class="hdlist1">
<strong>-version</strong> 
</dt>
<dd>
<p>
display version information and exit
</p>
</dd>
</dl></div>
</div>
</div>
<div class="sect1">
<h2 id="_reporting_bugs">REPORTING BUGS</h2>
<div class="sectionbody">
<div class="paragraph"><p>Report bugs to <a href="https://github.com/genometools/genometools/issues">https://github.com/genometools/genometools/issues</a>.</p></div>
</div>
</div>
<div id="footer">
Copyright &copy; 2007-2023 The <i>GenomeTools</i> authors.
</div>
</div>
</body>
</html>
