diff --git a/src/core/parseutils.c b/src/core/parseutils.c
index 8d512b8..d087419 100644
--- a/src/core/parseutils.c
+++ b/src/core/parseutils.c
@@ -96,7 +96,7 @@ int gt_parse_double(double *out, const char *nptr)
   dval = strtod(nptr, &ep);
   if (nptr[0] == '\0' || *ep != '\0')
     return -1;
-  if (errno == ERANGE && (dval == 0 || dval == HUGE_VAL || dval == -HUGE_VAL))
+  if (errno == ERANGE && (dval == 0 || dval == DBL_MAX || dval == -DBL_MIN))
     return -1;
   *out = dval;
   return 0;
diff --git a/src/external/lua-5.1.4/src/lmathlib.c b/src/external/lua-5.1.4/src/lmathlib.c
index 441fbf7..b464dd6 100644
--- a/src/external/lua-5.1.4/src/lmathlib.c
+++ b/src/external/lua-5.1.4/src/lmathlib.c
@@ -7,6 +7,7 @@
 
 #include <stdlib.h>
 #include <math.h>
+#include <float.h>
 
 #define lmathlib_c
 #define LUA_LIB
@@ -252,7 +253,7 @@ LUALIB_API int luaopen_math (lua_State *L) {
   luaL_register(L, LUA_MATHLIBNAME, mathlib);
   lua_pushnumber(L, PI);
   lua_setfield(L, -2, "pi");
-  lua_pushnumber(L, HUGE_VAL);
+  lua_pushnumber(L, DBL_MAX);
   lua_setfield(L, -2, "huge");
 #if defined(LUA_COMPAT_MOD)
   lua_getfield(L, -1, "fmod");
diff --git a/src/external/rnv-1.7.8/xsd.c b/src/external/rnv-1.7.8/xsd.c
index 14b022c..803c8e5 100644
--- a/src/external/rnv-1.7.8/xsd.c
+++ b/src/external/rnv-1.7.8/xsd.c
@@ -4,6 +4,7 @@
 #include <stdlib.h> /*atof,atol,strtol*/
 #include <string.h> /*strlen*/
 #include <math.h> /*HUGE_VAL*/
+#include <float.h>
 #include <assert.h>
 #include "u.h"
 #include "xmlc.h"
@@ -406,8 +407,8 @@ static int chkdec(struct facets *fp,char *s,int n) {
 }
 
 static double atodn(char *s,int n) {
-  return s_tokcmpn("-INF",s,n)==0?-HUGE_VAL
-    : s_tokcmpn("INF",s,n)==0?HUGE_VAL
+  return s_tokcmpn("-INF",s,n)==0?DBL_MIN
+    : s_tokcmpn("INF",s,n)==0?DBL_MAX
     : atof(s);
 }
 static double atod(char *s) {return atodn(s,strlen(s));}
